clustered/client/src/main/java/org/ehcache/clustered/client/internal/store/operations/BaseKeyValueOperation.java

```diff
@@ -16,41 +16,110 @@
+public abstract class BaseKeyValueOperation<K, V> implements Operation<K, V> {
 
-  public BaseKeyValueOperation(final K key, final V value) {
-    super(key);
+  private final K key;
+  private final V value;
+
+  BaseKeyValueOperation(K key, V value) {
+    if(key == null) {
+      throw new NullPointerException("Key can not be null");
+    }
+    if(value == null) {
+      throw new NullPointerException("Value can not be null");
+    }
+    this.key = key;
     this.value = value;
   }
 
+  BaseKeyValueOperation(ByteBuffer buffer, Serializer<K> keySerializer, Serializer<V> valueSerializer) {
+    OperationCode opCode = OperationCode.valueOf(buffer.get());
+    if (opCode != getOpCode()) {
+      throw new IllegalArgumentException("Invalid operation: " + opCode);
+    }
+    int keySize = buffer.getInt();
+    buffer.limit(buffer.position() + keySize);
+    ByteBuffer keyBlob = buffer.slice();
+    buffer.position(buffer.limit());
+    buffer.limit(buffer.capacity());
+    try {
+      this.key = keySerializer.read(keyBlob);
+      this.value = valueSerializer.read(buffer.slice());
+    } catch (ClassNotFoundException e) {
+      throw new CodecException(e);
+    }
+  }
+
+  public K getKey() {
+    return key;
+  }
+
   public V getValue() {
-    return this.value;
+    return value;
+  }
+
+  /**
+   * Here we need to encode two objects of unknown size: the key and the value.
+   * Encoding should be done in such a way that the key and value can be read
+   * separately while decoding the bytes.
+   * So the way it is done here is by writing the size of the payload along with
+   * the payload. That is, the size of the key payload is written before the key
+   * itself. The value payload is written after that.
+   *
+   * While decoding, the size is read first and then reading the same number of
+   * bytes will get you the key payload. Whatever that is left is the value payload.
+   */
+  @Override
+  public ByteBuffer encode(final Serializer<K> keySerializer, final Serializer<V> valueSerializer) {
+    ByteBuffer keyBuf = keySerializer.serialize(key);
+    ByteBuffer valueBuf = valueSerializer.serialize(value);
+
+    int size = BYTE_SIZE_BYTES +   // Operation type
+               INT_SIZE_BYTES +    // Size of the key payload
+               keyBuf.remaining() + // the kay payload itself
+               valueBuf.remaining();  // the value payload
+
+    ByteBuffer buffer = ByteBuffer.allocate(size);
+
+    buffer.put(getOpCode().getValue());
+    buffer.putInt(keyBuf.remaining());
+    buffer.put(keyBuf);
+    buffer.put(valueBuf);
+
+    buffer.flip();
+    return buffer;
   }
 
   @Override
   public String toString() {
-    return super.toString() + ", value: " + value;
+    return getOpCode() + "# key: " + key + ", value: " + value;
   }
 
   @Override
   public boolean equals(final Object obj) {
-    if(!super.equals(obj)) {
+    if(obj == null) {
       return false;
     }
     if(!(obj instanceof BaseKeyValueOperation)) {
```

Comment:
Although now that I think of it - do we need equals and hashCode for these types?